libbindgen/tests/tests.rs

```diff
@@ -0,0 +1,150 @@
+extern crate diff;
+extern crate env_logger;
+extern crate libbindgen;
+extern crate log;
+
+use std::env;
+use std::fs;
+use std::io::{BufRead, BufReader, Error, ErrorKind, Read};
+use std::path::{Path, PathBuf};
+
+fn spawn_bindgen(header: &PathBuf, builder: libbindgen::Builder) -> Result<(), Error> {
+    let file_name = header.file_name()
+        .expect("spawn_bindgen expects a file");
+
+    let mut expected = PathBuf::from(header);
+    expected.pop();
+    expected.pop();
+    expected.push("expectations");
+    expected.push("tests");
+    expected.push(file_name);
+    expected.set_extension("rs");
+
+    let output = match builder.generate() {
+        Ok(bindings) => bindings.to_string(),
+        Err(_) => "".to_string(),
+    };
+
+    let mut buffer = String::new();
+    let f = try!(fs::File::open(&expected));
+    let _ = try!(BufReader::new(f).read_to_string(&mut buffer));
+
+    if output == buffer {
+        Ok(())
+    } else {
+        println!("diff expected generated\n--- expected: {:?}\n+++ generated from: {:?}",
+            expected, header);
+        for diff in diff::lines(&buffer, &output) {
+            match diff {
+                diff::Result::Left(l) => println!("-{}", l),
+                diff::Result::Both(l, _) => println!(" {}", l),
+                diff::Result::Right(r) => println!("+{}", r),
+            }
+        }
+        Err(Error::new(ErrorKind::Other, "Header and binding differ!"))
+    }
+}
+
+fn create_bindgen_builder(header: &PathBuf) -> Result<libbindgen::Builder, Error> {
+    let mut builder = libbindgen::builder()
+        .header(header.to_str().unwrap())
+        .no_unstable_rust()
+        .raw_line("")
+        .raw_line("#![allow(non_snake_case)]")
+        .raw_line("");
+
+    let source = try!(fs::File::open(header));
+    let reader = BufReader::new(source);
+
+    let head: Vec<_> = try!(reader.lines().take(5).collect());
+    let flags: Vec<_> = head.iter()
+        .filter(|line| line.contains("bindgen-flags:"))
+        .flat_map(|line| line.split("bindgen-flags:").last())
+        .flat_map(|flags| flags.trim().split_whitespace())
+        .collect();
+
+    let mut it = flags.into_iter();
+    while let Some(flag) = it.next() {
+        match flag {
+            "--enable-cxx-namespaces" => {
```

Comment:
Agreed with @emilio that we should have a function in `libbindgen` that can be shared here and with the binary:

```rust
pub fn builder_from_flags<I>(iter: I) -> Builder
    where I: Iterator<&str>
{
    // ...
}
```